Support for multiple resource types and controller-runtime refactor

joelanford opened the pull request:
This PR is meant to fix #41 and #42.

It is based on new scaffolding produced by running the following command from a build of `operator-sdk`'s [`refactor/controller-runtime`](https://github.com/operator-framework/operator-sdk/tree/refactor/controller-runtime) branch:
 ```
$ operator-sdk new helm-app-operator
```

It also adds support for watching multiple resource types and managing releases from multiple helm charts using a new YAML file for configuration. Support for the existing environment variables is maintained if the configuration file is not defined/present.

Much of the design inspiration is taken from [`github.com/water-hole/ansible-operator`](https://github.com/water-hole/ansible-operator)

Other improvements were also made:
* Automatic support for in-cluster and out-of-cluster mode.
* Minor improvements to input validation. For example, checking to ensure helm chart directories are defined

hasbro17 commented on helm-app-operator/pkg/helm/controller/controller.go:
```diff
@@ -0,0 +1,68 @@
+	}
+
+	// Emit events once per minute regardless of detected changes.
+	r := newReconcileLoop(time.Minute*1, options.GVK, mgr.GetClient())
```
I'm wondering if we really need this (both here and in the ansible-operator).
@shawn-hurley Can you comment on why we can't just set the resync option for the manager(which sets it for the cache)? 
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L91-L95
```
mgr, err := manager.New(cfg, manager.Options{Namespace: namespace, SyncPeriod: time.Minute*1})
```

hasbro17 replied:
Discussed offline.
For our use case the manager's resync is ***almost*** equivalent to the reconcile loop.
https://github.com/kubernetes-sigs/controller-runtime/pull/88#issuecomment-408546638
So we can use either.

However the recently added [RequeueAfter](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/reconcile/reconcile.go#L30-L31) is probably the best way to periodically re-queue and re-trigger the reconcile for each CR.

joelanford replied:
Agreed. I'll make the change and verify expected behavior.

hasbro17 commented on helm-app-operator/pkg/helm/controller/reconcile.go:
```diff
@@ -0,0 +1,70 @@
+	_, ok := s.(map[string]interface{})
+	if !ok {
+		u.Object["spec"] = map[string]interface{}{}
+		r.Client.Update(context.TODO(), u)
```
Check return error

hasbro17 commented on helm-app-operator/pkg/helm/controller/reconcile.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	// Update the CR with the updated status.
+	r.Client.Update(context.TODO(), u)
```
Check return error.

hasbro17 commented on helm-app-operator/pkg/helm/installer/installer.go:
```diff
@@ -0,0 +1,348 @@
+	m := map[schema.GroupVersionKind]Installer{}
+
+	// Determine the watches file to use
+	watchesFile, ok := os.LookupEnv(HelmChartWatchesEnvVar)
```
The HelmChartWatchesEnvVar  could be set to empty, `HELM_CHART_WATCHES =""`.
In which case `ok=true` and `watchesFile=""`. We'll end up trying to load the watches file in that case right?

joelanford replied:
We would, in which case there would be an error returned by NewFromWatches. I figured that if the environment variable is set (regardless of its value), we should use it since the user explicitly set it.

I could go either way on this though. We could also fall back to the default file for the empty string case.

joelanford replied:
Added a commit make the logic in this section of code more obvious.

joelanford commented on helm-app-operator/pkg/helm/controller/reconcile.go:
```diff
@@ -0,0 +1,70 @@
+		if apierrors.IsNotFound(err) {
+			u, err = r.Installer.UninstallRelease(u)
+			if err != nil {
+				log.Print(err)
```
It probably makes the most sense to go with the outcome of operator-framework/operator-sdk#503, since there's discussion and voting there about the preferences and technical merits of various loggers (and just for the sake of consistency). I'll open a new issue to follow up once that decision is made.

joelanford replied:
New issue #44 to track logging updates.

hasbro17 commented:
It seems periodic reconciles will result in increasing release numbers for no changes to the CR.
https://github.com/helm/helm/issues/3177 
But we do need to periodically re-trigger reconciles to account for changes to unwatched(created) resources.

joelanford commented:
Agreed. I'll remove a check that skips upgrades for unchanged releases to allow unwatched resources to be reconciled. I'll investigate the ramifications of and possible workarounds for increasing helm release version numbers.

hasbro17 commented:
One way to ignore an event resulting from a CR status update is via the status subresource
https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#status-subresource

When the status subresource is enabled for a CRD then any change to `cr.spec` will increment the `metadata.generation` field. While changes to `cr.status` will not update the generation field. 

And we can filter out Update events that don't have the generation changed via a [predicate](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/predicate/predicate.go) in our controller:
```Go
pred := predicateForGenerationChange()
err = c.Watch(&source.Kind{Type: helmApp}, &handler.EnqueueRequestForObject{}) , pred)

...

func predicateForGenerationChange() predicate.Predicate {
	return predicate.Funcs{
		CreateFunc: func(e event.CreateEvent) bool {
			return true
		},
		UpdateFunc: func(e event.UpdateEvent) bool {
			// extract and compare metadata.generation between old and new object
			return generationNotEqual(e.objectOld, e.objectNew)
		},
		DeleteFunc: func(e event.DeleteEvent) bool {
			return true
		},
	}
}
```

The only problem with the above is that the subresource status isn't enabled by default for k8s 1.10. It's alpha and feature-gated. It's [beta](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#subresources) in k8s 1.11 and 1.12 and should be enabled by default.

For 1.10 we'll need to figure out another way besides the generation.

And for `kubectl -f apply` what kind of helper did you mean? 
Did you mean doing something like a `CreateOrUpdate()` on a `manifest.yaml`?

joelanford commented:
I think this latest commit solves the problem without the need for the status subresource, though that's not something I knew about, so thanks for the tip. 

For `kubectl apply`-like functionality, I went with a create or patch style reconciliation. I had first tried create or replace logic, but that caused various failures. For example `ClusterIP` is an immutable service field, so replacing the spec failed.

I think this raises a question about whether we should support helm's force option, and if so, how that would be conveyed in a CR (perhaps an annotation?). See https://github.com/helm/helm/issues/1958

joelanford commented:
Added another commit which fixes a few more bugs:

**Release retries**
If a helm release update fails, tiller stores the failed release in its storage backend with details about the failure. During the next reconciliation, we pull the latest release from the storage backend, not the latest _deployed_ release. 

Most likely, the latest release is the one that just failed and now we're retrying it, so the latest release and the candidate release are identical. In this case, we reconcile the underlying resources when we should be retrying the failed update.

**Fix:** use the latest _deployed_ release, not the latest release. 

**Failing uninstalls**
Now that we're using `reconcile.Result.RequeueAfter`, the reconciliation loop is executed once even after the watched object has been deleted. The uninstall logic is unable to locate the release associated with the now-deleted and already-uninstalled release, resulting in repeated reconciliation failures.

**Fix:** ignore missing release errors during uninstalls.

joelanford commented:
@hasbro17 - This should be ready for another look. 

The PR is getting pretty big and has now expanded beyond the original scope. I'm wondering if it should be split into multiple PRs?

There's probably still a bit more to do as well:
- Updates to docs and examples
- Repo layout updates, to more closely align with `ansible-operator`'s layout

@alecmerdler - Are there any community members that should be aware of these proposed changes?

hasbro17 commented:
@joelanford If you could break this PR up into 2 that would be great.

Perhaps a more logical split would be one for changes to the installer and another one for the new controller-runtime refactoring. But I'll let you decide how best to split it up.
You can keep this one open until then.

And yeah updating the docs and the project layout can be done in followup PRs after that.

joelanford commented:
Closing since this was split up among multiple PRs that have been merged (#45, #46, #47, #48, #51)
